mantle/common/counter.py

```diff
@@ -144,3 +146,41 @@ def UpDownCounter(n, cout=True, has_ce=False, has_reset=False, **kwargs):
               has_ce=has_ce, has_reset=has_reset)(**kwargs)
 
 
+def DefineCeilFloorUpDownCounter(m, has_ce=False, has_reset=False):
+    """
+    Counter between 0 and m - 1 that uses the minimum number of bits.
+    
+    This counter counts up by 1 if U is True and D is False, down by 1 if U is False and D is True,
+    no change if both are False or True
+    
+    Also, unlike normal updown counter, up by 1 has no effect if already at m - 1 and
+    down by 1 has nefect if already at 0.
+    
+    :param m: The value the counter counts up to
+    :param has_ce: Whether this counter should a clock-enable input
+    :return: An updown counter circuit
+    """
+    class CeilFloorUpDownCounter(Circuit):
+        num_bits = math.ceil(math.log(m, 2))
```

Comment:
In the future, you could also use m.bitutils.clog2 (since magma is usually imported already)